packages/design-system/src/components/Dialog/Dialog.tsx

```diff
@@ -115,22 +117,7 @@ export const Dialog = (props: DialogProps) => {
     containerRef.current?.focus();
   }, [containerRef]);
 
-  // Prevent scrolling the page behind the dialog. Needs to use useLayoutEffect
-  // because we need to grab the window scroll position before the dialog renders
-  // and messes it up.
-  useLayoutEffect(() => {
-    // https://css-tricks.com/prevent-page-scrolling-when-a-modal-is-open/
-    const y = window.scrollY ?? 0;
-    const bodyClass = 'ds--dialog-open';
-    document.body.classList.add(bodyClass);
-    document.body.style.setProperty('--body_top--dialog-open', `-${y}px`);
-    document.documentElement.style.setProperty('scroll-behavior', 'auto');
-    return () => {
-      document.body.classList.remove(bodyClass);
-      window.scrollTo({ top: y, behavior: 'auto' });
```

Comment:
Making the state dependent on the value of `isOpen` instead of following when it's mounted and unmounted, it meant that it was no longer as simple as a _use effect_ and _cleanup_. Because I no longer had direct access to the `y` value created in the top of the _use effect_ in a closure, I had to start tracking that state separately. I didn't love that, so you'll notice that I ended up moving this all into a separate module and used a reducer, which I ended up trying to model as a finite state machine. It might be a little more code than is needed, but I feel more confident in it being a finite state machine, especially since I've been able to leverage TypeScript to help make sure my logic is correct.